File: data_environment/sources/groupprivate.1.cpp
/*
* @@name:       groupprivate.1
* @@type:       C++ 
* @@operation:  run 
* @@expect:     success
* @@version:    omp_6.0
*/
#include <omp.h>
#include <stdio.h>

void init(int *x, int n, int tid)
{
   #pragma omp for
   for (int i = 0; i < n; i++)
     x[i] = tid + i;

}

void foo(int &sum, int tid)
{
   static int x[100];
   #pragma omp groupprivate(x) 
  
   init(x, 100, tid);
   
   #pragma omp for reduction(+:sum)
   for (int i = 0; i < 100; i++) {
     sum += x[i];
   }   
}
#pragma omp declare_target enter(foo)

int main()
{
   int sums[4] = {0,0,0,0};

   #pragma omp target teams num_teams(4) thread_limit(100)
   #pragma omp parallel
   foo(sums[omp_get_team_num()], omp_get_team_num());

   if( sums[0] != 4950 || sums[1] != 5050 ||
       sums[2] != 5150 || sums[3] != 5250 ){
     printf("FAILED\n");
     return 1;
   }   
   printf("PASSED\n");

   return 0;
}

File: data_environment/sources/groupprivate.1.f90
! @@name:       groupprivate.1
! @@type:       F-free
! @@operation:  run
! @@expect:     success
! @@version:    omp_6.0
module mfunc
contains
  subroutine init(x, n, tid)
    implicit none
    integer, intent(in)  :: tid, n
    integer, intent(out) :: x(n)
    integer              :: i
    
    ! Initialize the array with the thread number
    !$omp do
    do i = 1, n
      x(i) = tid + i
    end do
  end subroutine init

  subroutine foo(sum, tid)
    implicit none
    integer, intent(inout) :: sum
    integer, intent(in)    :: tid
    integer                :: i
    integer, save          :: x(100)
    !$omp groupprivate(x)
    !$omp declare_target 

    call init(x,100,tid)

    ! Perform the reduction operation
    !$omp do reduction(+:sum) 
    do i = 1, 100
      sum = sum + x(i)
    end do
  end subroutine foo

end module mfunc

program main
  use mfunc
  use omp_lib
  implicit none

  integer :: sums(4) = (/ 0, 0, 0, 0 /)
  integer :: team_num, thread_num

  !$omp target teams num_teams(4) thread_limit(100)
  !$omp parallel private(team_num, thread_num)
    team_num = omp_get_team_num()
    thread_num = omp_get_thread_num()
    call foo(sums(team_num+1), team_num)
  !$omp end parallel
  !$omp end target teams

  if( sums(1) /= 5050 .or. sums(2) /= 5150 .or. &
      sums(3) /= 5250 .or. sums(4) /= 5350 ) then
    print*, "FAILED"
    stop 1
  endif
  print *, "PASSED"
end program
